bson-codec-design/DESIGN.md

```diff
@@ -0,0 +1,177 @@
+themselves. For example, users who generate types using a protocol buffer library
+or users who want to handle the processing of types in the standard library.
+
+The design of the v1 encoder and decoder for the Go driver presents problems when users wish to
+have more control over how specific types are handled when encoding or decoding. The control that
+users require exists on two main levels:
+- control over how individual types are encoded or decoded
+- control over how tags are processed
+
+The granularity of control users require can be defined in three levels:
+- Control over how individual types are encoded and decoded when they are embedded in another type,
+such as a map or a struct
+- Control over how individual types are encoded and decoded when they are provided directly to the
+encoder or decoder
+- Control over how tags are processed and the parameters that are returned
+
+The proposed design allows for the first two granularities to be handled together.
+
+## High Level Design
+### Registries
+To enable users to control how types are encoded and decoded, the proposed design uses a type
+registry. This registry will be provided in any context where encoding or decoding occurs. The registry
+will be used to lookup a type's codec, which will then be used for the actual encoding a decoding.
+This will allow users to control nearly all aspects of the encoding and decoding process. One
+failing of the v1 design is that users cannot control the encoding and decoding of types
+they do not own, e.g. `time.Time`. The proposed design handles this by allowing users to register a
+codec for any type.
+
+Pointer and value types are handled separately in the registry as this simplifies the registry and
+the implementations of codecs, since they will only have to worry about a single type and not the
+type plus a pointer to that type. While likely not useful, there is no design
+reason to prevent users from registering codecs for internal types like `bson.Element`,
+`bson.Document`, or `bson.Reader`. Users can override the functionality of built in types as
+well. The registry will be used for all encoding and decoding operations, since the cost of calling
+`reflect.TypeOf` is inexpensive.
+
+There is a single type, `Registry`, that contains an instance of a `TypeRegistry` and
+`InterfaceRegistry`. The `Registry` type is used to find the required codec during codec processing.
+The `Registry` type will also have a generic codec registry, which only contains two codecs, one for
+structs and one for maps. These codecs are only used if both the `TypeRegistry` and the
+`InterfaceRegistry` do not return a codec. The user can override these generic codecs.
+
+#### No Global Registries
+The proposed design does not contain a global type nor a global interface registry. This is mainly
+because [package level variables are generally discouraged in
+Go](http://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html). Instead, registries are
```

Comment:
meh... there are always exceptions for everything, as mentioned at the bottom. I believe this is one of them. And this one, in particular, is not difficult to reason about.

Reply:
I wouldn't say it's not difficult to reason about. I think any level of `init` function magical registering of stuff is difficult to reason about, especially since it requires you to load up Go Guru every time you want to figure out all of the things that are registered.